man/objdump-ng.1
.TH OBJDUMP-NG "1" "October 2026" "objtools 1.10" "User Commands"
.SH NAME
objdump-ng: display information about ELF and PE files.

.SH SYNOPSIS
objdump-ng [\fB\-H\fR|\fB\-\-help\fR]
           [\fB\-V\fR|\fB\-\-version\fR]
           [\fB\-v\fR|\fB\-\-verbose\fR]
           [\fB\-D\fR|\fB\-\-disassemble\-all\fR]
           [\fB\-C\fR|\fB\-\-capstone\-all\fR]
           [\fB\-S\fR|\fB\-\-source\fR]
           [\fB\-T\fR|\fB\-\-dynamic\-symbols\fR]
           [\fB\-R\fR|\fB\-\-dynamic\-reloc\fR]
           [\fB\-d\fR|\fB\-\-disassemble\fR]
           [\fB\-c\fR|\fB\-\-capstone\fR]
           [\fB\-g\fR|\fB\-\-debugging\fR]
           [\fB\-e\fR|\fB\-\-debugging\-tags\fR]
           [\fB\-C\fR|\fB\-\-demangle\fR]
           [\fB\-s\fR|\fB\-\-full\-contents\fR]
           [\fB\-t\fR|\fB\-\-symbols\fR]
           [\fB\-r\fR|\fB\-\-reloc\fR]
           [\fB\-f\fR|\fB\-\-file\-headers\fR]
           [\fB\-p\fR|\fB\-\-private\-headers\fR]
           [\fB\-h\fR|\fB\-\-section\-headers\fR]
           [\fB\-x\fR|\fB\-\-all\-headers\fR]
           [\fB\-l\fR|\fB\-\-line\-numbers\fR]
           [\fB\-\-\-no\-addresses\fR]
           [\fB\-\-\-no\-show\-raw\-insn\fR]
           [\fB\-\-\-prefix\-addresses\fR]
           [\fB\-\-\-start\-address=\fR<address>]
           [\fB\-\-\-stop\-address=\fR<address>]
           [\fB\-\-\-capstone\fR]
           [\fB\-X\fR|\fB\-\-hash\fR]
           [\fB\-W[aAcFikKLmrRsptTuUgX]\fR|
            \fB\-\-dwarf\fR[=abbrev,=addr,=cu_index,=frames\-interp,=info,=links,=follow\-links,
                    =decodedline,=macro,=aranges,=ranges,=str,=pubnames,=pubtypes,=trace_aranges,
                    =trace_abbrev,=trace_info,=gdb_index,=enhanced]]
           [\fB\-M\fR <options>|\fB\-\-disassembler\-options=\fR<options>]
           \fIobjfile\fR...

.SH DESCRIPTION
\fBobjdump-ng\fR displays information about one or more object files. The options control what particular information to display. This information is mostly useful to programmers who are working on the compilation tools, as opposed to programmers who just want their program to compile and work.

\fIobjfile...\fR are the object files to be examined. When you specify archives, it shows information on each of the member object files.

.SH OPTIONS
.IP "\fB\-H\fR" 4
.IX Item "-H"
.PD 0
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
.PD
Print a summary of the options to objdump-ng and exit.
.IP "\fB\-V\fR" 4
.IX Item "-V"
.PD 0
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
.PD
Print the version number of objdump-ng and exit.
.IP "\fB\-v\fR" 4
.IX Item "-v"
.PD 0
.IP "\fB\-\-verbose\fR" 4
.IX Item "--verbose"
.PD
Print additional information.
.IP "\fB\-D\fR" 4
.IX Item "-D"
.PD 0
.IP "\fB\-\-disassemble\-all\fR" 4
.IX Item "--disassemble-all"
.PD
Like \fB-d\fR, but disassemble the contents of all sections, not just those expected to contain instructions.
Equivalent to specifying: \fB\-\-disassemble\fR.
.IP "\fB\-C\fR" 4
.IX Item "-C"
.PD 0
.IP "\fB\-\-capstone\-all\fR" 4
.IX Item "--capstone-all"
.PD
Like \fB-c\fR, but disassemble the contents of all sections, not just those expected to contain instructions.
Equivalent to specifying: \fB\-\-disassemble\fR.
.IP "\fB\-S\fR" 4
.IX Item "-S"
.PD 0
.IP "\fB\-\-source\fR" 4
.IX Item "--source"
.PD
Display source code intermixed with disassembly, if possible. Implies \fB\-d\fR.
Equivalent to specifying: \fB\-\-disassemble\fR.
.IP "\fB\-T\fR" 4
.IX Item "-T"
.PD 0
.IP "\fB\-\-dynamic\-symbols\fR" 4
.IX Item "--dynamic-symbols"
.PD
Print the dynamic symbol table entries of the file.
.IP "\fB\-R\fR" 4
.IX Item "-R"
.PD 0
.IP "\fB\-\-dynamic\-reloc\fR" 4
.IX Item "--dynamic-reloc"
.PD
Print the dynamic relocation entries of the file. This is only meaningful for dynamic objects, such as certain types of shared libraries.
.IP "\fB\-d\fR" 4
.IX Item "-d"
.PD 0
.IP "\fB\-\-disassemble\fR" 4
.IX Item "--disassemble"
.PD
Display the assembler mnemonics for the machine instructions from the input file.
.IP "\fB\-c\fR" 4
.IX Item "-c"
.PD 0
.IP "\fB\-\-capstone\fR" 4
.IX Item "--capstone"
.PD
Display the assembler mnemonics for the machine instructions from the input file, use the Capstone Disassembler Framework.
Equivalent to specifying: \fB\-\-disassemble\fR.
.IP "\fB\-g\fR" 4
.IX Item "-g"
.PD 0
.IP "\fB\-\-debugging\fR" 4
.IX Item "--debugging"
.PD
Display debugging information.
.IP "\fB\-e\fR" 4
.IX Item "-e"
.PD 0
.IP "\fB\-\-debugging\-tags\fR" 4
.IX Item "--debugging-tags"
.PD
Like \fB\-g\fR, but the information is generated in a format compatible with ctags tool.
.IP "\fB\-s\fR" 4
.IX Item "-s"
.PD 0
.IP "\fB\-\-full\-contents\fR" 4
.IX Item "--full-contents"
.PD
Display the full contents of any sections requested.
.IP "\fB\-t\fR" 4
.IX Item "-t"
.PD 0
.IP "\fB\-\-symbols\fR" 4
.IX Item "--symbols"
.PD
Print the symbol table entries of the file.
.IP "\fB\-r\fR" 4
.IX Item "-r"
.PD 0
.IP "\fB\-\-reloc\fR" 4
.IX Item "--reloc"
.PD
Print the relocation entries of the file.
.IP "\fB\-f\fR" 4
.IX Item "-f"
.PD 0
.IP "\fB\-\-file\-headers\fR" 4
.IX Item "--file-headers"
.PD
Display summary information from the overall header of each of the objfile files.
.IP "\fB\-P\fR" 4
.IX Item "-P"
.PD 0
.IP "\fB\-\-private\-headers\fR" 4
.IX Item "--private-headers"
.PD
Print information that is specific to the object file format.
.IP "\fB\-h\fR" 4
.IX Item "-h"
.PD 0
.IP "\fB\-\-section\-headers\fR" 4
.IX Item "--section-headers"
.PD
Display summary information from the section headers of the object file.
.IP "\fB\-x\fR" 4
.IX Item "-x"
.PD 0
.IP "\fB\-\-all\-headers\fR" 4
.IX Item "--all-headers"
.PD
Display all available header information, including the symbol table and relocation entries.
Equivalent to specifying: \fB\-\-symbols\fR, \fB\-\-reloc\fR, \fB\-\-file\-headers\fR, \fB\-\-private\-headers\fR and \fB\-\-section\-headers\fR.
.IP "\fB\-l\fR" 4
.IX Item "-l"
.PD 0
.IP "\fB\-\-line\-numbers\fR" 4
.IX Item "--line-numbers"
.PD
Label the display (using debugging information) with the filename and source line numbers corresponding to the object code or relocs shown. Only useful with \fB\-d\fR, \fB\-D\fR, or \fB\-r\fR.
.IP "\fB\-\-no\-addresses\fR" 4
.IX Item "--no-addresses"
.PD
When disassembling, don't print addresses on each line or for symbols and relocation offsets. In combination with \fB\-\-no\-show\-raw\-insn\fR this may be useful for comparing compiler output.
.IP "\fB\-\-no\-show\-raw\-insn\fR" 4
.IX Item "--no-show-raw-insn"
.PD
When disassembling instructions, do not print the instruction bytes. This is the default when \fB\-\-prefix\-addresses\fR is used.
.IP "\fB\-\-prefix\-addresses\fR" 4
.IX Item "--prefix-addresses"
.PD
When disassembling, print the complete address on each line.
.IP "\fB\-\-start\-address=<address>\fR" 4
.IX Item "--start-address=<address>"
.PD
Start displaying data at the specified address. This affects the output of the \fB\-d\fR, \fB\-r\fR and \fB\-s\fR options.
.IP "\fB\-\-stop\-address=<address>\fR" 4
.IX Item "--stop-address=<address>"
.PD
Stop displaying data at the specified address. This affects the output of the \fB\-d\fR, \fB\-r\fR and \fB\-s\fR options.
.IP "\fB\-\-capstone\fR" 4
.IX Item "--capstone"
.PD
When disassembling, use the Capstone Disassembler Framework.
.IP "\fB\-X\fR" 4
.IX Item "-X"
.PD 0
.IP "\fB\-\-hash\fR" 4
.IX Item "--hash"
.PD
Calculates and displays the SHA-256 hash.
.IP "\fB\-w[aAcFikKLmrRsptTuUgX]\fR" 4
.IX Item "-w[aAcFikKLmrRsptTuUgX]"
.PD 0
.IP "\fB\-\-dwarf[=abbrev,=addr,=cu_index,=frames-interp,=info,=links,=follow-links,=decodedline,=macro,=aranges,=ranges,=str,=pubnames,=pubtypes,=trace_aranges,=trace_abbrev,=trace_info,=gdb_index,=enhanced]\fR" 4
.IX Item "--dwarf[=abbrev,=addr,=cu_index,=frames-interp,=info,=links,=follow-links,=decodedline,=macro,=aranges,=ranges,=str,=pubnames,=pubtypes,=trace_aranges,=trace_abbrev,=trace_info,=gdb_index,=enhanced]"
.PD
Displays the contents of the \s-1DWARF\s0 debug sections in the file, if any are present.
.RS 4
.ie n .IP """a""" 4
.IX Item "a"
.PD 0
.ie n .IP """=abbrev""" 4
.IX Item "=abbrev"
.PD
Displays the contents of the \fB.debug_abbrev\fR section.
.ie n .IP """A""" 4
.IX Item "A"
.PD 0
.ie n .IP """=addr""" 4
.IX Item "=addr"
.PD
Displays the contents of the \fB.debug_addr\fR section.
.ie n .IP """c""" 4
.IX Item "c"
.PD 0
.ie n .IP """=cu_index""" 4
.IX Item "=cu_index"
.PD
Displays the contents of the \fB.debug_cu_index\fR|\fB.debug_tu_index\fR sections.
.ie n .IP """F""" 4
.IX Item "F"
.PD 0
.ie n .IP """=frames\-interp""" 4
.IX Item "=frames-interp"
.PD
Display the interpreted contents of a \fB.debug_frame\fR section.
.ie n .IP """i""" 4
.IX Item "i"
.PD 0
.ie n .IP """=info""" 4
.IX Item "=info"
.PD
Displays the contents of the \fB.debug_info\fR section.
.ie n .IP """k""" 4
.IX Item "k"
.PD 0
.ie n .IP """=links""" 4
.IX Item "=links"
.PD
Displays the contents of the \fB.gnu_debuglink\fR|\fB.gnu_debugaltlink\fR sections.
.ie n .IP """K""" 4
.IX Item "K"
.PD 0
.ie n .IP """=follow\-links""" 4
.IX Item "=follow-links"
.PD
Display the contents of any selected debug sections that are found in linked, separate debug info file(s).
.ie n .IP """L""" 4
.IX Item "L"
.PD 0
.ie n .IP """=decodedline""" 4
.IX Item "=decodedline"
.PD
Displays the interpreted contents of the \fB.debug_line\fR section.
.ie n .IP """m""" 4
.IX Item "m"
.PD 0
.ie n .IP """=macro""" 4
.IX Item "=macro"
.PD
Displays the contents of the \fB.debug_macro\fR|\fB.debug_macinfo\fR sections.
.ie n .IP """r""" 4
.IX Item "r"
.PD 0
.ie n .IP """=aranges""" 4
.IX Item "=aranges"
.PD
Displays the contents of the \fB.debug_aranges\fR section.
.ie n .IP """R""" 4
.IX Item "R"
.PD 0
.ie n .IP """=ranges""" 4
.IX Item "=ranges"
.PD
Displays the contents of the \fB.debug_ranges\fR|\fB.debug_rnglists\fR sections.
.ie n .IP """s""" 4
.IX Item "s"
.PD 0
.ie n .IP """=str""" 4
.IX Item "=str"
.PD
Displays the contents of the \fB.debug_str\fR|\fB.debug_line_str\fR|\fB.debug_str_offsets\fR sections.
.ie n .IP """p""" 4
.IX Item "p"
.PD 0
.ie n .IP """=pubnames""" 4
.IX Item "=pubnames"
.PD
Displays the contents of the \fB.debug_pubnames\fR|\fB.debug_gnu_pubnames\fR sections.
.ie n .IP """t""" 4
.IX Item "t"
.PD 0
.ie n .IP """=pubtypes""" 4
.IX Item "=pubtypes"
.PD
Displays the contents of the \fB.debug_pubtypes\fR|\fB.debug_gnu_pubtypes\fR sections.
.ie n .IP """T""" 4
.IX Item "T"
.PD 0
.ie n .IP """=trace_aranges""" 4
.IX Item "=trace_aranges"
.PD
Displays the contents of the \fB.trace_aranges\fR section.
.ie n .IP """u""" 4
.IX Item "u"
.PD 0
.ie n .IP """=trace_abbrev""" 4
.IX Item "=trace_abbrev"
.PD
Displays the contents of the \fB.trace_abbrev\fR section.
.ie n .IP """U""" 4
.IX Item "U"
.PD 0
.ie n .IP """=trace_info""" 4
.IX Item "=trace_info"
.PD
Displays the contents of the \fB.trace_info\fR section.
.ie n .IP """g""" 4
.IX Item "g"
.PD 0
.ie n .IP """=gdb_index""" 4
.IX Item "=gdb_index"
.PD
Displays the contents of the \fB.gdb_index\fR|\fB.debug_names\fR sections.
.ie n .IP """X""" 4
.IX Item "X"
.PD 0
.ie n .IP """=enhanced""" 4
.IX Item "=enhanced"
.PD
Displays the contents in the enhanced mode format.
.RE
.IP "\fB\-M <options>\fR" 4
.IX Item "-M <options>"
.PD 0
.IP "\fB\-\-disassembler\-options=<options>\fR" 4
.IX Item "--disassembler-options=<options>"
.PD
Pass target specific information to the disassembler.
.RS 4
.ie n .IP """att""" 4
.IX Item "att"
.PD
Select the AT&T syntax mode.
.ie n .IP """intel""" 4
.IX Item "intel"
.PD
Select the Intel syntax mode.
.ie n .IP """arm32""" 4
.IX Item "arm32"
.PD 0
.ie n .IP """arm64""" 4
.IX Item "arm64"
.ie n .IP """i386""" 4
.IX Item "i386"
.ie n .IP """i8086""" 4
.IX Item "i8086"
.ie n .IP """x86-64""" 4
.IX Item "x86-64"
.ie n .IP """riscv32""" 4
.IX Item "riscv32"
.ie n .IP """riscv64""" 4
.IX Item "riscv64"
.ie n .IP """mips1""" 4
.IX Item "mips1"
.ie n .IP """mips2""" 4
.IX Item "mips2"
.ie n .IP """mips3""" 4
.IX Item "mips3"
.ie n .IP """mips4""" 4
.IX Item "mips4"
.ie n .IP """mips5""" 4
.IX Item "mips5"
.ie n .IP """mips32""" 4
.IX Item "mips32"
.ie n .IP """mips32r2""" 4
.IX Item "mips32r2"
.ie n .IP """mips32r6""" 4
.IX Item "mips32r6"
.ie n .IP """mips64""" 4
.IX Item "mips64"
.ie n .IP """mips64r2""" 4
.IX Item "mips64r2"
.ie n .IP """mips64r6""" 4
.IX Item "mips64r6"
.PD
Select disassembly for the given architecture.
.ie n .IP """amd64""" 4
.IX Item "amd64"
.PD 0
.ie n .IP """intel64""" 4
.IX Item "intel64"
.PD
Select between AMD64 ISA and Intel64 ISA.
.ie n .IP """addr16""" 4
.IX Item "addr16"
.PD 0
.ie n .IP """addr32""" 4
.IX Item "addr32"
.ie n .IP """addr64""" 4
.IX Item "addr64"
.ie n .IP """data16""" 4
.IX Item "data16"
.ie n .IP """data32""" 4
.IX Item "data32"
.PD
Specify the default address size and operand size.
.RE

.SH SEE ALSO
detect-ng(1), convert-ng(1), coredump-ng(1), objcopy-ng(1), objdump-ng(1), objdwarf-ng(1), objhash-ng(1), readelf-ng(1), readpe-ng(1)

.SH COPYRIGHT
.IX Header "COPYRIGHT"
MIT License
.PP
Copyright (c) 2025 Kyle Wickens

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.



CHANGELOG
v1.2
-----------------------------------------------------------------------
[+] Added big-endian support.
[+] Added MIPS architecture support.
[+] Improved RISC-V architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################

v1.1
-----------------------------------------------------------------------
[+] Added ARM Thumb architecture support.
[+] Added RISC-V architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################

v1.0
-----------------------------------------------------------------------
[+] Added ARM architecture support.
[+] Improved x86-64 architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################



src/objdump-ng.c
#include "opcode.h"
#include "printf.h"
#include "objdump.h"
#include "options.h"

#include "static/objdump-ng.ci"

#define THIS_NAME "objdump-ng"

static int usage(poptions_t o, const char* name, const args_t args[], const int ecode) {
  int n = 0;
  n += usage_name(o, name, args, zDESCRIPTION);
  n += usage_synopsis0(o, name, args);
  n += usage_synopsis1(o, name, zDEBUGELFARGS, zOBJDUMPARGS0, zOBJDUMPARGS1);
  n += usage_synopsis2(o, name, zDISASSEMBLEARGS, zOBJDUMPARGS2, zOBJDUMPARGS3);
  n += usage_objname(o, name, "objfile");
  n += printf_eol();
  n += usage_description(o, name, args, zOBJDESCRIPTION);
  n += usage_options0(o, name, args);
  n += usage_options1(o, name, zDEBUGELFARGS, zOBJDUMPARGS0, zOBJDUMPARGS1);
  n += usage_options2(o, name, zDISASSEMBLEARGS, zOBJDUMPARGS2, zOBJDUMPARGS3);
  n += usage_seealso(o, name, args);
  n += usage_copyright(o, name, args);

  return ecode;
}

static int get_options_objdump(poptions_t o, int argc, char** argv, char* name) {
  if (0 == argc) {
    return usage(o, THIS_NAME, zOBJDUMPARGS, ECODE_ARGUMENTS);
  }

  strname(o->prgname, name);
  printf_errname(THIS_NAME);

  for (int i = 0; i < argc; ++i) {
    if ('-' == argv[i][0] && '-' == argv[i][1]) {
      MALLOCA(char, arg0, 256);
      MALLOCA(char, arg1, 256);

      if (ECODE_ISOK(breakup_args(argv[i], arg0, NELEMENTS(arg0), arg1, NELEMENTS(arg1)))) {
        if (0 == xstrcmp(arg0, zOBJDUMPARGS1)) {
          imode_t ocdump = get_options2(o, zDEBUGELFARGS, arg1);
          if (0 == ocdump) {
            return odeath(o, THIS_NAME, arg1);
          }
          o->ocdump |= ocdump;
        } else if (0 == xstrcmp(arg0, zOBJDUMPARGS3)) {
          imode_t ocdump = get_options2(o, zDISASSEMBLEARGS, arg1);
          if (0 == ocdump) {
            return odeath(o, THIS_NAME, arg1);
          }
          o->ocdump |= ocdump;
        } else if (0 == xstrcmp(arg0, "--start-address")) {
          o->saddress = atovalue(arg1);
        } else if (0 == xstrcmp(arg0, "--stop-address")) {
          o->eaddress = atovalue(arg1);
        } else {
          return odeath(o, THIS_NAME, arg0 + 2);
        }
      } else {
        imode_t action = get_options2(o, zOBJDUMPARGS, argv[i]);
        if (0 == action) {
          return odeath(o, THIS_NAME, argv[i] + 2);
        }
        o->action |= action;
      }
    } else if ('-' == argv[i][0]) {
      if (argv[i][0] == zOBJDUMPARGS0[0] && argv[i][1] == zOBJDUMPARGS0[1]) {
        imode_t ocdump = get_options1(o, zDEBUGELFARGS, argv[i] + 1);
        if (0 == ocdump && argv[i][2]) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        o->ocdump |= ocdump ? ocdump : set_options1(o, zDEBUGELFARGS);
      } else if (0 == xstrcmp(argv[i], zOBJDUMPARGS2)) {
        imode_t ocdump = get_options2(o, zDISASSEMBLEARGS, argv[++i]);
        if (0 == ocdump) {
          return odeath(o, THIS_NAME, argv[i]);
        }
        o->ocdump |= ocdump;
      } else {
        imode_t action = get_options1(o, zOBJDUMPARGS, argv[i]);
        if (0 == action) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        o->action |= action;
      }
    } else if (0 == o->inpname[0]) {
      xstrncpy(o->inpname, argv[i], NELEMENTS(o->inpname));
    } else {
      return odeath(o, THIS_NAME, argv[i]);
    }
  }

  o->ocdump |= get_ocdump(o, zOBJDUMPSWAP, o->action);

  if (o->action & OPTPROGRAM_VERSION) {
    return version0(o, THIS_NAME, zOBJDUMPARGS);
  }

  if (o->action & OPTPROGRAM_HELP) {
    return usage(o, THIS_NAME, zOBJDUMPARGS, ECODE_OK);
  }

  return ECODE_OK;
}

int main(int argc, char* argv[]) {
  int x = ECODE_MALLOC;
  poptions_t o = omalloc();
  if (o) {
    x = get_options_objdump(o, argc - 1, argv + 1, argv[0]);
    if (ECODE_ISOK(x) && o->inpname[0]) {
      occonfig(o->prgname, "x86_64-pc-linux-gnu");

      handle_t p = ocopen(o->inpname);
      if (p) {
        x = objdump(p, o);
      } else {
        printf_e("'%s': no such file.", o->inpname);
      }

      occlose(p);
    }

    ofree(o);
  }

  return x;
}

